sender.go

```diff
@@ -108,6 +112,11 @@ func (s *Sender) Close(ctx context.Context) error {
 	s.clientMu.Lock()
 	defer s.clientMu.Unlock()
 
+	return s.close(ctx)
+}
+
+// closes the connection.  callers *must* hold the client write lock before calling!
+func (s *Sender) close(ctx context.Context) error {
```

Comment:
Why is it necessary to move the functionality in Close() to close()?

Reply:
So that in `Recover()` we can `close()` and `newSessionAndLink()` while holding the write lock.

Reply:
In the previous implementation there was a window between `Close()` and `newSessionAndLink()` where the write lock wasn't held.  So you could `Close()` the connection then `Send()` would acquire the read lock however the connection hadn't been rebuilt yet.